source/common/stream_info/utility.cc

```diff
@@ -5,180 +5,44 @@
-const std::string ResponseFlagUtils::RATE_LIMITED = "RL";
-const std::string ResponseFlagUtils::UNAUTHORIZED_EXTERNAL_SERVICE = "UAEX";
-const std::string ResponseFlagUtils::RATELIMIT_SERVICE_ERROR = "RLSE";
-const std::string ResponseFlagUtils::STREAM_IDLE_TIMEOUT = "SI";
-const std::string ResponseFlagUtils::INVALID_ENVOY_REQUEST_HEADERS = "IH";
-const std::string ResponseFlagUtils::DOWNSTREAM_PROTOCOL_ERROR = "DPE";
-const std::string ResponseFlagUtils::UPSTREAM_MAX_STREAM_DURATION_REACHED = "UMSDR";
-const std::string ResponseFlagUtils::RESPONSE_FROM_CACHE_FILTER = "RFCF";
-const std::string ResponseFlagUtils::NO_FILTER_CONFIG_FOUND = "NFCF";
-const std::string ResponseFlagUtils::DURATION_TIMEOUT = "DT";
-const std::string ResponseFlagUtils::UPSTREAM_PROTOCOL_ERROR = "UPE";
-const std::string ResponseFlagUtils::NO_CLUSTER_FOUND = "NC";
-
-void ResponseFlagUtils::appendString(std::string& result, const std::string& append) {
+void ResponseFlagUtils::appendString(std::string& result, absl::string_view append) {
   if (result.empty()) {
     result = append;
   } else {
-    result += "," + append;
+    result += ",";
+    result += append;
   }
 }
 
 const std::string ResponseFlagUtils::toShortString(const StreamInfo& stream_info) {
   std::string result;
 
-  static_assert(ResponseFlag::LastFlag == 0x1000000, "A flag has been added. Fix this code.");
-
-  if (stream_info.hasResponseFlag(ResponseFlag::FailedLocalHealthCheck)) {
-    appendString(result, FAILED_LOCAL_HEALTH_CHECK);
-  }
-
-  if (stream_info.hasResponseFlag(ResponseFlag::NoHealthyUpstream)) {
-    appendString(result, NO_HEALTHY_UPSTREAM);
-  }
-
-  if (stream_info.hasResponseFlag(ResponseFlag::UpstreamRequestTimeout)) {
-    appendString(result, UPSTREAM_REQUEST_TIMEOUT);
-  }
-
-  if (stream_info.hasResponseFlag(ResponseFlag::LocalReset)) {
-    appendString(result, LOCAL_RESET);
-  }
-
-  if (stream_info.hasResponseFlag(ResponseFlag::UpstreamRemoteReset)) {
-    appendString(result, UPSTREAM_REMOTE_RESET);
-  }
-
-  if (stream_info.hasResponseFlag(ResponseFlag::UpstreamConnectionFailure)) {
-    appendString(result, UPSTREAM_CONNECTION_FAILURE);
-  }
-
-  if (stream_info.hasResponseFlag(ResponseFlag::UpstreamConnectionTermination)) {
-    appendString(result, UPSTREAM_CONNECTION_TERMINATION);
-  }
-
-  if (stream_info.hasResponseFlag(ResponseFlag::UpstreamOverflow)) {
-    appendString(result, UPSTREAM_OVERFLOW);
-  }
-
-  if (stream_info.hasResponseFlag(ResponseFlag::NoRouteFound)) {
-    appendString(result, NO_ROUTE_FOUND);
-  }
-
-  if (stream_info.hasResponseFlag(ResponseFlag::DelayInjected)) {
-    appendString(result, DELAY_INJECTED);
-  }
-
-  if (stream_info.hasResponseFlag(ResponseFlag::FaultInjected)) {
-    appendString(result, FAULT_INJECTED);
-  }
-
-  if (stream_info.hasResponseFlag(ResponseFlag::RateLimited)) {
-    appendString(result, RATE_LIMITED);
+  for (const auto& [flag_string, flag] : ALL_RESPONSE_STRING_FLAGS) {
```

Comment:
actually the above comment is superfluous. What you really want to do is make a temp `vector<string_view>` and then join that:
```
std::vector<absl::string_view> flag_strings;
for (const auto& [flag_string, flag] : ALL_RESPONSE_STRING_FLAGS) {
  if (stream_info.hasResponseFlag(flag)) {
    flag_strings.push_back(flag_string);
  }
}
if (flag_strings.empty()) {
  return NONE;
}
return absl::StrJoin(flag_strings, ",");
```
This will result in a lot less string resizing and appending.

TBH I'm not sure if this is called enough to warrant optimization but the optimized version is not too complex.

Reply:
That's awesome! I replace the std::vector by InlinedVector and copy-paste each character below.